### DISCUSSION TITLE
[RFC] Button Semantic DOM Structure

### DISCUSSION THREAD

**Discussion by MadCcc** [ant-design/ant-design#41392] [RFCs - archive]
## Summary

1. Add wrapper to icon node of Button.
2. Add `classNames` and `styles` to Button. 

## API

| Property | Description | 
| --- | --- |
| icon | The `icon` node. |

## Basic Example

```tsx
import React from 'react';
import { Button } from 'antd';

export default () => (
  <Button classNames={{ icon: 'custom-icon' }} />
)
```

### Why we need a wrapper for icon node?
Currently we do not have a `-icon` class and wrapper for icon node, and the margin between icon and text was implemented only for `anticon`.
So we need to apply a wrapper for icon node to make all icons passed by user would get correct gap with text. And the semantic dom mentioned above refers to this wrapper.

## How we teach this

In **feature** branch:

Add document for Button semantic dom structure.

**Comment by arvinxx** [comment]
loading 好像是和 icon 并列存在的？ 需要给 loading 添加一个 wrapper 么

**Reply by MadCcc** [reply]
loading 在实现上就是替换了 icon 的，和 icon 是一体的。

**Comment by afc163** [comment]
只有 icon 感觉意义不大，icon 也可以单独设 className 的：`icon={<XxxIcon className="xx" />}`

**Reply by MadCcc** [reply]
这个 className 是给到 icon 外层的。目前 Button 的 icon 插槽缺少一个 wrapper，导致自定义的 icon 和 text 之间没有预设的间距，需要手动调整。这个 RFC 另一个目的就是给 Button 的 icon 加一层 wrapper。

**Reply by afc163** [reply]
按钮间距这个问题最好用 https://github.com/ant-design/ant-design/issues/32952#issuecomment-987352843 这种方案更一劳永逸一点。

**Reply by MadCcc** [reply]
> 按钮间距这个问题最好用 [#32952 (comment)](https://github.com/ant-design/ant-design/issues/32952#issuecomment-987352843) 这种方案更一劳永逸一点。

这个方案影响会更大，比如在 Button 的 children 中写了两个 span，就会受这个影响中间多了间距，但可能并不是用户的本意。
这种写法不用 `inline-flex` 也能实现，类似于：
```css
.ant-btn > *:not(:first-child) {
  margin-left: 8px;
}
```
